Refactor Localization

hanny-ph opened the pull request:
This MR basically does the following:

1. Remove `lookupFDashLocalizations` usage from all layers (Model and View/Widget Layer).
   * This will make faiadashu's localizations can be overridden from host project
1. Remove Locale from every widget.
   * Doing this will simplify localization by centralizing selected Locale in MaterialApp, or can be overridden by wrapping a widget using `Localizations.override`.
1. Update example app.

fjwong commented on lib/questionnaires/view/src/xhtml.dart:
```diff
@@ -54,8 +55,14 @@ class Xhtml extends StatelessWidget {
 
-    final xhtml = renderingString.xhtmlText;
-    final plainText = renderingString.plainText;
+    final xhtml = renderingString.xhtmlText.translate(
+      renderingString.extensions,
+      Localizations.localeOf(context),
+    );
+    final plainText = renderingString.plainText.translate(
+      renderingString.extensions,
+      Localizations.localeOf(context),
+    );
```
I'm not sure if this is correct, since the FHIR `extensions` are added to the `text` element of questions, and may not contain any actual xhtml markup. Won't this end up replacing the html markup with a translated string instead?

I also see that down below we're calling `_extractBase64FromImgTag(xhtml);` which sounds like it may break too, although I'm not sure.

hanny-ph replied:
Thank you @fjwong, yeah you are correct, xhtml should not be translated, but please correct me if I am wrong, `plaintText` might still be translated since RenderingString can be initialized by `Text`.

fjwong replied:
@hanny-ph I think `plainText` may or may not need translation, but I'm wondering if this would be the correct place to do so. Please refer to https://github.com/sujrd/faiadashu/pull/49#discussion_r1347205071.

fjwong reviewed (commented):
@hanny-ph Thank you for the changes. I really appreciate the effort.

I've checked the code briefly and left you a few comments. I'll try to conduct another more in-depth review later.

Just a question, I see that we've largely replaced returning validation messages in favor of throwing and catching exceptions instead which appears to complicate things a bit. Is there any particular advantage in doing this?

I'm also a bit concerned, as this changes the behavior quite drastically, since exceptions basically stop the regular execution flow of the app, so we'll need to make absolutely sure that we're still running code that is supposed to run after a validation takes place.

hanny-ph commented:
> Just a question, I see that we've largely replaced returning validation messages in favor of throwing and catching exceptions instead which appears to complicate things a bit. Is there any particular advantage in doing this?

The main reason for this change is to make Model layer doesn't have any reference to any component from widget/view layer.
Currently, `FDashLocalizations` is heavily used by the models during validation, e.g. `validatorRequiredItem`, `validatorEntryFormat`.

To me, Localization belongs to View Layer and this also aligns with Flutter nature that needs `BuildContext` to get `Localization` based on the selected `Locale`.

Doing this, will also remove `lookupFDashLocalizations` use.

Another route, we can pass around `FDashLocalizations` object from View to Model layer.

> I'm also a bit concerned, as this changes the behavior quite drastically, since exceptions basically stop the regular execution flow of the app, so we'll need to make absolutely sure that we're still running code that is supposed to run after a validation takes place.

Yeah, you are correct. I found several incorrect implementations when addressing the comments.

hanny-ph commented on lib/questionnaires/view/item/answer/src/coding_answer_filler.dart:
```diff
@@ -43,13 +43,20 @@ class _CodingInputControl extends AnswerInputControl<CodingAnswerModel> {
+              : null,
+          errorText: errorText,
+        )
+      ],
```
Addressed [here](https://github.com/sujrd/faiadashu/pull/49/commits/c98aee0b2e108635a4f225fec9066132c3f0d175).

hanny-ph commented on lib/questionnaires/model/item/src/question_item_model.dart:
```diff
@@ -430,8 +419,9 @@ class QuestionItemModel extends ResponseItemModel {
-          (ex is FhirPathEvaluationException) ? ex.message : ex.toString();
+      // errorText =
+      //     (ex is FhirPathEvaluationException) ? ex.message : ex.toString();
+      if (ex is FhirPathEvaluationException) {}
```
Sorry, it was not intended, and has been fixed [here](https://github.com/sujrd/faiadashu/pull/49/commits/c98aee0b2e108635a4f225fec9066132c3f0d175). Thanks!

hanny-ph commented on example/lib/questionnaire_launch_tile.dart:
```diff
@@ -58,11 +58,13 @@ class _QuestionnaireLaunchTileState extends State<QuestionnaireLaunchTile> {
       fhirResourceProvider: _questionnaireProvider,
       launchContext: widget.launchContext,
-      locale: _locale,
+      locale: Locale('ja'),
```
Sorry, it was not intended, and has been fixed [here](https://github.com/sujrd/faiadashu/pull/49/commits/c98aee0b2e108635a4f225fec9066132c3f0d175). Thanks!

hanny-ph commented on example/lib/main.dart:
```diff
@@ -358,7 +401,7 @@ class _HomePageState extends State<HomePage> {
                           fhirResourceProvider: AssetResourceProvider.singleton(
                             questionnaireResourceUri,
-                            'assets/instruments/framingham-hcdc.json',
+                            'assets/instruments/triage.json',
```
Updated, Thanks!

hanny-ph commented on lib/questionnaires/model/item/answer/src/numerical_answer_model.dart:
```diff
@@ -259,7 +265,8 @@ class NumericalAnswerModel extends AnswerModel<String, Quantity> {
   /// * Keeps the unit
   Quantity? copyWithTextInput(String textInput) {
-    final valid = validateInput(textInput) == null;
+    final valid = (() => validateInput(textInput)).callSafely() ?? false;
```
Since we no longer use throwing and catching exceptions, it's been removed. Thanks

fjwong commented on lib/questionnaires/model/item/answer/src/coding_answer_option_model.dart:
```diff
@@ -132,7 +132,7 @@ class CodingAnswerOptionModel {
 
     final optionPrefix = (plainOptionPrefix != null)
-        ? RenderingString.fromText(plainOptionPrefix)
+        ? RenderingString.fromText(plainOptionPrefix, extensions: extensions)
```
I'm not sure if we should pass the `extensions` here or not. From what I see, these should affect the question option itself, but here we are setting them to `plainOptionPrefix`, which is by itself coming from one of the `extensions` too.

hanny-ph replied:
This is also been addressed [here](https://github.com/sujrd/faiadashu/pull/49/commits/3598028e169d969b943d32e8b739ed21aebcca11). Thanks!

fjwong commented on lib/questionnaires/model/item/answer/src/coding_answer_option_model.dart:
```diff
@@ -154,6 +154,7 @@ class CodingAnswerOptionModel {
         forDisplay = _createForDisplay(coding, locale, questionnaireItemModel);
         optionText = RenderingString.fromText(
           plainText,
+          extensions: extensions,
```
I'm also not sure about this one either. Here `plainText` is not coming directly from the question text itself, but from whatever is being returned by `_createMultiColumn` in line 153, so we may end up replacing the actual content when the text is translated.

hanny-ph replied:
It's been addressed [here](https://github.com/sujrd/faiadashu/pull/49/commits/3598028e169d969b943d32e8b739ed21aebcca11), thanks!

hanny-ph commented on lib/questionnaires/view/item/answer/src/numerical_answer_filler.dart:
```diff
@@ -222,11 +226,14 @@ class _NumberFieldInputControl
+              return answerModel
+                  .validateInput(inputValue)
+                  ?.getMessage(FDashLocalizations.of(context));
+            },
```
Thanks for the comment!

I addressed here: https://github.com/sujrd/faiadashu/pull/49/commits/47debb7ee4dbaf646adac98ebf9e8ec26256cf75

hanny-ph commented on lib/questionnaires/model/item/src/response_item_model.dart:
```diff
@@ -99,6 +95,11 @@ abstract class ResponseItemModel extends FillerItemModel {
+        nodeUid,
+        questionnaireItemModel.constraintHuman,
+      );
+    }
```
Sorry, I overlooked this part, but it's been addressed [here](https://github.com/sujrd/faiadashu/pull/49/commits/0962199475ed76c89a402f3516ddce83471e7980).

hanny-ph commented:
@hanny-ph To answer [this comment](https://github.com/sujrd/faiadashu/pull/49#discussion_r1347205071), [this comment](https://github.com/sujrd/faiadashu/pull/49#discussion_r1347060247), [this comment](https://github.com/sujrd/faiadashu/pull/49#discussion_r1347056389) and [this comment](https://github.com/sujrd/faiadashu/pull/49#discussion_r1347223978).

I made several adjustments:

1. Move the translation logic in RenderingString during object creation.
1. Replace questionnaire response model and its listeners when locale is changed.
    * Without this, translation will not work when locale is changed.
    
I apologize for the big changes. Please continue a review when you have time.

fjwong commented on lib/questionnaires/view/src/questionnaire_stepper_page_view.dart:
```diff
@@ -30,8 +30,9 @@ class QuestionnaireStepperPageView extends StatefulWidget {
-  PageController _pageController = PageController();
+class _QuestionnaireStepperPageViewState
+    extends State<QuestionnaireStepperPageView> {
+  PageController _pageController = PageController(keepPage: true);
```
Just for curiosity, why is it that we need `keepPage: true`?

hanny-ph replied:
Actually, if we go with the current approach, this is not needed anymore.

Basically, this will retain the position of page view when the screen is reloaded after locale is changed.

hanny-ph replied:
Removed [here](https://github.com/sujrd/faiadashu/pull/49/commits/33aae122977430f71d8f89271e0a1f560a518856). thanks!

fjwong commented on lib/questionnaires/model/src/rendering_string.dart:
```diff
@@ -44,6 +48,7 @@ class RenderingString with Diagnosticable {
     this.renderingStyle,
     this.renderingXhtml,
     this.renderingMarkdown,
+    this.extensions,
```
I'm wondering if we still need to keep the extensions as a public property of RenderingString. I see were not using renderingString.extensions anymore.

hanny-ph replied:
Thanks for the reminder, removed [here](https://github.com/sujrd/faiadashu/pull/49/commits/0f82ef052054f94c403434c34b87bc37010ec48d).

fjwong commented on lib/questionnaires/model/item/answer/src/coding_answer_option_model.dart:
```diff
@@ -132,7 +133,11 @@ class CodingAnswerOptionModel {
+        ? RenderingString.fromText(
+            plainOptionPrefix,
+            locale: locale,
+            extensions: extensions,
```
Just leaving this comment here so that we don't forget to change this.

hanny-ph replied:
Thanks for the comment, addressed [here](https://github.com/sujrd/faiadashu/pull/49/commits/3598028e169d969b943d32e8b739ed21aebcca11).

fjwong reviewed (approved):
@hanny-ph Thank you very much for the changes.

I think this is good to go.
